File: Baekjoon/Class2/32.10828 스택.java
import java.io.*;
import java.util.EmptyStackException;
import java.util.StringTokenizer;

class Main {
    private static class Stack<T> { // 연결 리스트 스택
        private static class Node<T> {  // 연결 리스트 노드
            T data;
            Node<T> next;
            public Node(T data, Node<T> next) {
                this.data = data;
                this.next = next;
            }
        }
        private Node<T> top; // 첫 노드
        private int size;

        public Stack() {
            top = null;
            size = 0;
        }
        public void push(T item) {
            size++;
            Node<T> nodeToPush = new Node<T>(item, top);
            top = nodeToPush;
        }
        public T pop() {
            if (empty()) throw new EmptyStackException();
            size--;
            T item = top.data;
            top = top.next;
            return item;
        }
        public int size() {
            return size;
        }
        public boolean empty() {
            return size == 0;
        }
        public T top() {
            if (empty()) throw new EmptyStackException();
            return top.data;
        }
    }

    public static void main(String[] args) throws IOException {
        int n;
        Stack<Integer> nums = new Stack<>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        for(int i = 0; i < n; i++) {
            String command;
            st = new StringTokenizer(br.readLine());
            command = st.nextToken();
            try {
                switch (command) {
                    case "push":
                        nums.push(Integer.parseInt(st.nextToken()));
                        break;
                    case "pop":
                        sb.append(nums.pop()).append('\n');
                        break;
                    case "size":
                        sb.append(nums.size()).append('\n');
                        break;
                    case "empty":
                        sb.append(nums.empty() ? 1 : 0).append('\n');
                        break;
                    case "top":
                        sb.append(nums.top()).append('\n');
                        break;
                }
            }
            catch (EmptyStackException e) { // 스택 빔
                sb.append("-1\n");
            }
        }
        System.out.print(sb);
        br.close();
    }
}

/* 라이브러리 사용
import java.io.*;
import java.util.*;

class Main {
    public static void main(String[] args) throws IOException {
        int n;
        Deque<Integer> nums = new ArrayDeque<>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        for(int i = 0; i < n; i++) {
            String command;
            st = new StringTokenizer(br.readLine());
            command = st.nextToken();
            switch (command) {
                case "push":
                    nums.push(Integer.parseInt(st.nextToken()));
                    break;
                case "pop":
                    if(nums.isEmpty()) {
                        sb.append("-1\n");
                        continue;
                    }
                    sb.append(nums.pop()).append('\n');
                    break;
                case "size":
                    sb.append(nums.size()).append('\n');
                    break;
                case "empty":
                    sb.append(nums.isEmpty() ? 1 : 0).append('\n');
                    break;
                case "top":
                    if(nums.isEmpty()) {
                        sb.append("-1\n");
                        continue;
                    }
                    sb.append(nums.peek()).append('\n');
                    break;
            }
        }
        System.out.print(sb);
        br.close();
    }
}
*/

File: Baekjoon/Class2/32.10828 스택.cpp
#include <iostream>
#include <string>
using namespace std;

template <typename T> // 연결 리스트 스택
class Stack {
private:
    struct Node { // 연결 리스트 노드
        T data;
        Node* next;
    };
    Node* _top; // 첫 노드
    int _size;
public:
    Stack() : _top(nullptr), _size(0) {}
    ~Stack() {
        while (!empty())
            pop();
    }
    void push(const T& data) {
        _size++;
        Node* node_to_push = new Node;
        node_to_push->data = data;
        node_to_push->next = _top;
        _top = node_to_push;
    }
    T pop() {
        if (empty()) throw out_of_range("stack is empty");
        _size--;
        Node* node_to_pop = _top;
        T data = node_to_pop->data;
        _top = node_to_pop->next;
        delete node_to_pop;
        return data;
    }
    int size() {
        return _size;
    }
    bool empty() {
        return _size == 0;
    }
    T top() {
        if (empty()) throw out_of_range("stack is empty");
        return _top->data;
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n;
    Stack<int> nums;

    cin >> n;
    while (n--) {
        string command;
        cin >> command;
        try {
            if (command == "push") {
                int num;
                cin >> num;
                nums.push(num);
            }
            else if (command == "pop")
                cout << nums.pop() << '\n';
            else if (command == "size")
                cout << nums.size() << '\n';
            else if (command == "empty")
                cout << nums.empty() << '\n';
            else if (command == "top")
                cout << nums.top() << '\n';
        }
        catch (const out_of_range& e) { // 스택 빔
            cout << -1 << '\n';
        }
    }
    return 0;
}

/* 라이브러리 사용
#include <iostream>
#include <string>
#include <stack>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n;
    stack<int> nums;

    cin >> n;
    while (n--) {
        string command;
        cin >> command;
        if (command == "push") {
            int num;
            cin >> num;
            nums.push(num);
        }
        else if (command == "pop") {
            if (nums.empty()) {
                cout << "-1\n";
                continue;
            }
            cout << nums.top() << '\n';
            nums.pop();
        }
        else if (command == "size")
            cout << nums.size() << '\n';
        else if (command == "empty")
            cout << nums.empty() << '\n';
        else if (command == "top") {
            if (nums.empty()) {
                cout << "-1\n";
                continue;
            }
            cout << nums.top() << '\n';
        }
    }
    return 0;
}
*/